### ISSUE TITLE
index_reader should evict the local index cache after skipping

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#12378]
Following up on #12271, the long reactor stalls seen there indicates lots of stuff is accumulating in the index_reader _local_index_cache, and since it's private to the index_reader instance we evict from it only on close().

Since the index "cursor" can only moev forward, never backwards, any index page can be evicted once fully processed since it won't be read again.  A good point to do that is after skipping forward, as in the case of using the index during cleanup compaction where we skip frequently over tokens the node doesn't own anymore.

**Comment by bhalevy** [comment]
Cc @tgrabiec @michoecho

**Comment by michoecho** [comment]
Yes, the fact that the cache is not cleared until the read end pretty much means that BYPASS CACHE isn't quite doing it's job — a scan may still pollute the cache. I noticed the problem some time ago, but AFAIK it was like that for many years, so I didn't act on it.

By the way, how did the cleanup in #12271 accumulate enough data in the local cache to cause a 800ms stall? Shouldn't it read the index only twice (for range start and range end) per vnode? ~500 index pages shouldn't need 800ms to be freed.

**Comment by tgrabiec** [comment]
> Yes, the fact that the cache is not cleared until the read end pretty much means that BYPASS CACHE isn't quite doing it's job — a scan may still pollute the cache. I noticed the problem some time ago, but AFAIK it was like that for many years, so I didn't act on it.

Normally reads only need one or two entries, so there's not much pollution. All readers started via CQL are like that. So it is doing its job for most reads.

> 
> By the way, how did the cleanup in #12271 accumulate enough data in the local cache to cause a 800ms stall? Shouldn't it read the index only ~twice (for range start and range end) per vnode? ~500 index pages shouldn't need 800ms to be freed.


Here we have a special reader used by cleanup operation which uses the `next_partition()` API to filter partitions, which uses index to skip.

**Comment by michoecho** [comment]
> Here we have a special reader used by cleanup operation which uses the next_partition() API to filter partitions, which uses index to skip.

Wait. I imagined that cleanup works by skipping to all owned ranges within the SSTable and copying each range to the new SSTable. Does it not do that, and instead it iterates over everything and checks *every* partition key for ownership separately?

**Comment by tgrabiec** [comment]
I see this in the code:

```
    cleanup_compaction(table_state& table_s, compaction_descriptor descriptor, compaction_data& cdata, compaction_type_options::upgrade opts)
        : cleanup_compaction(table_s, std::move(descriptor), cdata, std::move(opts.owned_ranges)) {}

    flat_mutation_reader_v2 make_sstable_reader() const override {
        return make_filtering_reader(regular_compaction::make_sstable_reader(), make_partition_filter());
    }

```

and in `filtering_reader`:

```
          if (!_filter(dk)) {
              return _rd.next_partition();
          }
```

I recall that it used fast_forward_to() at some point, maybe it changed.

**Comment by tgrabiec** [comment]
next_partition() is inefficient if the skipped ranges are wide, needless to say.

**Comment by tgrabiec** [comment]
If it changed it changed a long time ago (prior to 2016). Maybe it was always like this.

**Comment by michoecho** [comment]
The entire implementation of cleanup sounds very inefficient then. Cleanup is something that could be performed without parsing the data file at all!

**Comment by tgrabiec** [comment]
Yes. This would be improved even further with compaction groups and tablets, where you need at most one index read when losing range.

**Comment by tgrabiec** [comment]
Or even could be lazy, if we implement partition range tombstones.

**Comment by tgrabiec** [comment]
>  Yes. This would be improved even further with compaction groups and tablets, where you need at most one index read when losing range.

Well, when splitting a tablet. When losing a tablet, cleanup is instantenous.

**Comment by michoecho** [comment]
Making cleanup lazy doesn't sound like a good idea. Preventing data resurrection is only one of cleanup's goals — the other one is to actually free disk space.

**Comment by tgrabiec** [comment]
It does if you can't run cleanup to completion and you need to decommission a node quickly because AWS gave you a deadline before the node will be taken down. We had a prod issue like that.

**Comment by tgrabiec** [comment]
decommission cannot be performed before cleanup completes on all nodes.

**Comment by bhalevy** [comment]
With tablets, adding or decommissioning nodes shouldn't necrssarily require tablet's splitting or merging, just migration across nodes to redistribute them (that without object storage, with it it'd be even simpler to transfer ownership of the objects across nodes).

So cleanup will be as easy as deleting the source tablet's files after it has been migrated.

**Comment by bhalevy** [comment]
For merging after a split, I don't see how a partition/token range tombstone could help delete any data that wouldcbe resurrected if it was deleted after the split and the respective tombstone got purged.

However if a merge brought the data back into the spurce partition before that tombstone was applied and merged, the "cleanup tombstone" would delete both the data prior to the split and the data after the merge if we don't have a way to distinguish between them, like a generation number similarcto the "token metadata generation" I prevoiously proposed, that could qualify the token.

**Comment by tgrabiec** [comment]
> For merging after a split, I don't see how a partition/token range tombstone could help delete any data that wouldcbe resurrected if it was deleted after the split and the respective tombstone got purged.

The partition tombstone would be created when splitting, not when merging. It would create an sstable which is logically equivalent to a compacted one. It would be applied only to that sstable, and you cannot add new data to an sstable, so there is no problem with it covering later writes to the database.

**Comment by bhalevy** [comment]
> > For merging after a split, I don't see how a partition/token range tombstone could help delete any data that wouldcbe resurrected if it was deleted after the split and the respective tombstone got purged.
> 
> The partition tombstone would be created when splitting, not when merging. It would create an sstable which is logically equivalent to a compacted one. It would be applied only to that sstable, and you cannot add new data to an sstable, so there is no problem with it covering later writes to the database.

I understand how that cpuld work on the logical level but I don't know what mechanics you had in mind to implement this. Woule there be one partition tombstone per tablet? Or one per sstable in the tablet at the time of splitting? Where is it stored? Is it a new component (the sstable components are immutable today) what does it contain?

**Comment by tgrabiec** [comment]
On Wed, Dec 21, 2022 at 8:45 PM Benny Halevy ***@***.***>
wrote:

> For merging after a split, I don't see how a partition/token range
> tombstone could help delete any data that wouldcbe resurrected if it was
> deleted after the split and the respective tombstone got purged.
>
> The partition tombstone would be created when splitting, not when merging.
> It would create an sstable which is logically equivalent to a compacted
> one. It would be applied only to that sstable, and you cannot add new data
> to an sstable, so there is no problem with it covering later writes to the
> database.
>
> I understand how that cpuld work on the logical level but I don't know
> what mechanics you had in mind to implement this. Woule there be one
> partition tombstone per tablet? Or one per sstable in the tablet at the
> time of splitting? Where is it stored? Is it a new component (the sstable
> components are immutable today) what does it contain?
>

It could be implemented on the sstable level, maybe as a metadata
component. It can be independent of tablets, it could be useful even today
as a way to implement "nodetool cleanup --fast". It was a rough idea, I
didn't think it through.

**Comment by bhalevy** [comment]
> It could be implemented on the sstable level, maybe as a metadata component. It can be independent of tablets, it could be useful even today as a way to implement "nodetool cleanup --fast". It was a rough idea, I didn't think it through.

Yes, the component could store an array of token range tombstones for all the ranges that were moved away from the compaction group.  This would work both for vnodes and for tablets.

**Comment by raphaelsc** [comment]
> > Here we have a special reader used by cleanup operation which uses the next_partition() API to filter partitions, which uses index to skip.
> 
> Wait. I imagined that cleanup works by skipping to all owned ranges within the SSTable and copying each range to the new SSTable. Does it not do that, and instead it iterates over everything and checks _every_ partition key for ownership separately?

Cleanup is deserializing in order to compact and expire, so bypassing data file is not possible

**Comment by michoecho** [comment]
> Cleanup is deserializing in order to compact and expire, so bypassing data file is not possible

Why does cleanup have to compact and expire? Isn't its job to remove data which is no longer replicated by the node?

**Comment by raphaelsc** [comment]
> > Cleanup is deserializing in order to compact and expire, so bypassing data file is not possible
> 
> Why does cleanup have to compact and expire? Isn't its job to remove data which is no longer replicated by the node?

during topology changes, it could happen that every sstable loses ownership of tokens. think of L0 sstables that spans roughly the entire token space. so we take advantage that we're reading sstables and writing their content, to compact and expire. Why not? Reduces write amplification. One thing that Tomek suggested is that we come up with a new filtering reader that skips to the next owned range, rather than calling next_partition() incrementally till the next owned token is reached.

**Comment by michoecho** [comment]
> Why not? Reduces write amplification.

Assuming you could avoid deserializing the data file altogether (I'm not sure if that's feasible, e.g. because then you couldn't compute the statistics file accurately), this is a tradeoff between CPU and disk throughput. You are reducing write amplification, but increasing CPU work. Should it be obvious to me that this is a good tradeoff? I don't have a feel for this.
